module/User/config/module.config.php

```diff
@@ -102,21 +165,31 @@
         // This is not documented somewhere, but more specific classes need to go first
         'factories' => [
             ApiPrincipalAdapter::class => ApiPrincipalAdapterFactory::class,
+            ApiPrincipalService::class => ApiPrincipalServiceFactory::class,
             ApiAuthenticationService::class => ApiAuthenticationServiceFactory::class,
             UserService::class => UserServiceFactory::class,
             UserMapper::class => UserMapperFactory::class,
             PasswordInterface::class => PasswordFactory::class,
             AuthenticationService::class => AuthenticationServiceFactory::class,
             ApiPrincipalMapper::class => ApiPrincipalMapperFactory::class,
+            ApiPrincipalForm::class => static function (ContainerInterface $container) {
+                $form = new ApiPrincipalForm($container->get(MvcTranslator::class));
+                $hydrator = $container->get(ClassMethodsHydrator::class);
+                $form->setHydrator($hydrator);
+
+                return $form;
+            },
         ],
         'invokables' => [
+            Login::class => Login::class,
             UserCreate::class => UserCreate::class,
             UserEdit::class => UserEdit::class,
-            Login::class => Login::class,
+            ClassMethodsHydrator::class => ClassMethodsHydrator::class,
```

Comment:
`new ClassMethodsHydrator()` in the factory is preferred.

Reply:
b48bbb2e9966b14482d8bd2b22ee043892724aac